### PR TITLE
Allow setting key/value in a metadata stored in the topology

### PR DESCRIPTION
**Pull request by kalfonso**
This PR allows setting key/value metadata in the topology leveraging "set" & "show" SQL statements.

Signed-off-by: Karel Alfonso Sague <kalfonso@squareup.com>

### CODE REVIEW

**Comment by kalfonso** [comment]
@sougou @demmer @tirsen this is an initial cut of using "SET" statement to manage metadata values. Looking for feedback on this. I've commented in the PR about alternative options to manage topology backed metadata

File: go/vt/vtgate/executor.go [hunk 1]
```diff
@@ -229,7 +229,7 @@ func (e *Executor) execute(ctx context.Context, safeSession *SafeSession, sql st
 	case sqlparser.StmtRollback:
 		return e.handleRollback(ctx, safeSession, sql, bindVars, logStats)
 	case sqlparser.StmtSet:
-		return e.handleSet(ctx, safeSession, sql, bindVars, logStats)
+		return e.handleSet(ctx, safeSession, sql, bindVars, destKeyspace, logStats)
```
Comment (tirsen) [inline go/vt/vtgate/executor.go]: I think automagically picking up the keyspace from the current target is a bit surprising. Existing uses of the `SET` statement aren't contextual like that.

Comment (kalfonso) [reply go/vt/vtgate/executor.go]: Done

File: go/vt/topo/metadata.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+// GetMetadata retrieves all metadata value that matches the given key regular expression. If empty all values are returned.
+func (ts *Server) GetMetadata(ctx context.Context, keyFilter string) (map[string]string, error) {
+	keys, err := ts.globalCell.ListDir(ctx, MetadataPath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	keyPattern := strings.Replace(regexp.QuoteMeta(keyFilter), "%", ".*", -1)
+	keyPattern = fmt.Sprintf("^%s$", keyPattern)
+	re := regexp.MustCompile(keyPattern)
```
Comment (alecthomas) [inline go/vt/topo/metadata.go]: I'd add a little comment at the end like `// Can never fail.`

File: go/vt/vtgate/executor.go [hunk 3]
```diff
@@ -465,6 +465,11 @@ func (e *Executor) handleSet(ctx context.Context, safeSession *SafeSession, sql
 		if k.Scope == sqlparser.GlobalStr {
 			return &sqltypes.Result{}, vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "unsupported in set: global")
 		}
+
+		if k.Scope == sqlparser.VitessMetadataStr {
```
Comment (alecthomas) [inline go/vt/vtgate/executor.go]: Use a `switch` here now that you have >1 case?

File: go/vt/vtgate/executor_test.go [hunk 4]
```diff
@@ -468,6 +468,69 @@ func TestExecutorSet(t *testing.T) {
...
+
+	set := "set @@vitess_metadata.app_keyspace_v1= '1'"
+	_, err := executor.Execute(context.Background(), "TestExecute", session, set, nil)
+	if err != nil {
+		t.Errorf("%s error: %v", set, err)
+	}
+
+	show := "show vitess_metadata variables like 'app_keyspace%'"
+	result, err := executor.Execute(context.Background(), "TestExecute", session, show, nil)
+	if err != nil {
```
Comment (alecthomas) [inline go/vt/vtgate/executor_test.go]: Is [testify](https://godoc.org/github.com/stretchr/testify/assert) not used in Vitess? :(

File: go/vt/vtgate/executor.go [hunk 5]
```diff
@@ -645,6 +650,56 @@ func (e *Executor) handleSet(ctx context.Context, safeSession *SafeSession, sql
...
+			return nil, err
+		}
+	} else {
+		metadata, err = ts.GetMetadata(ctx, opt.Filter.Like)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	rows := make([][]sqltypes.Value, 0, 2)
```
Comment (alecthomas) [inline go/vt/vtgate/executor.go]: I'd use `len(metadata)` as the capacity here.

File: go/vt/vtgate/executor_test.go [hunk 6]
```diff
@@ -468,6 +468,69 @@ func TestExecutorSet(t *testing.T) {
...
+			buildVarCharRow("app_keyspace_v2", "2"),
+		},
+		RowsAffected: 2,
+	}
+
+	if !reflect.DeepEqual(gotqr, wantqr) {
+		t.Errorf("want migrations %v, result %v", wantqr, gotqr)
+	}
+
+	show = "show vitess_metadata variables"
```
Comment (tirsen) [inline go/vt/vtgate/executor_test.go]: oh I didn't know the `variables` keyword. I assume this is "standard"?

File: go/vt/vtgate/executor.go [hunk 7]
```diff
@@ -680,6 +735,10 @@ func (e *Executor) handleShow(ctx context.Context, safeSession *SafeSession, sql
 
 	switch show.Type {
 	case sqlparser.KeywordString(sqlparser.COLLATION), sqlparser.KeywordString(sqlparser.VARIABLES):
+		if show.Scope == sqlparser.VitessMetadataStr {
+			return e.handleShowVitessMetadata(ctx, safeSession, destKeyspace, show.ShowTablesOpt)
```
Comment (tirsen) [inline go/vt/vtgate/executor.go]: You're still passing in the keyspace here?

File: go/vt/vtgate/executor.go [hunk 8]
```diff
@@ -645,6 +650,56 @@ func (e *Executor) handleSet(ctx context.Context, safeSession *SafeSession, sql
...
+	}
+
+	if err := ts.UpsertMetadata(ctx, k.Key, val); err != nil {
+		return nil, err
+	}
+
+	return &sqltypes.Result{}, nil
+}
+
+func (e *Executor) handleShowVitessMetadata(ctx context.Context, session *SafeSession, ksName string, opt *sqlparser.ShowTablesOpt) (*sqltypes.Result, error) {
```
Comment (tirsen) [inline go/vt/vtgate/executor.go]: I think you're not using `ksName`?

File: go/vt/vtgate/executor.go (hunk 5 above)
Comment (kalfonso) [reply go/vt/vtgate/executor.go]: Done

File: go/vt/vtgate/executor.go (hunk 8 above)
Comment (kalfonso) [reply go/vt/vtgate/executor.go]: Removed

File: go/vt/vtgate/executor.go (hunk 3 above)
Comment (kalfonso) [reply go/vt/vtgate/executor.go]: Done

File: go/vt/topo/metadata.go (hunk 2 above)
Comment (kalfonso) [reply go/vt/topo/metadata.go]: Done

File: go/vt/vtgate/executor.go (hunk 7 above)
Comment (kalfonso) [reply go/vt/vtgate/executor.go]: Removed

File: go/vt/vtgate/executor_test.go (hunk 6 above)
Comment (kalfonso) [reply go/vt/vtgate/executor_test.go]: Yes, wanted to keep same syntax

```
SHOW [GLOBAL | SESSION] VARIABLES
    [LIKE 'pattern' | WHERE expr]
```
https://dev.mysql.com/doc/refman/8.0/en/show-variables.html

File: go/vt/vtgate/executor_test.go (hunk 4 above)
Comment (kalfonso) [reply go/vt/vtgate/executor_test.go]: It seems to be used only in one test go/vt/vtgate/engine/ordered_aggregate_test.go

File: go/vt/vtgate/executor.go [hunk 9]
```diff
@@ -645,6 +649,63 @@ func (e *Executor) handleSet(ctx context.Context, safeSession *SafeSession, sql
 	return &sqltypes.Result{}, nil
 }
 
+func (e *Executor) handleSetVitessMetadata(ctx context.Context, session *SafeSession, k sqlparser.SetKey, v interface{}) (*sqltypes.Result, error) {
+	//TODO(kalfonso): move to its own acl check and consolidate into an acl component that can handle multiple operations (vschema, metadata)
+	allowed := vschemaacl.Authorized(callerid.ImmediateCallerIDFromContext(ctx))
```
Comment (kalfonso) [inline go/vt/vtgate/executor.go]: @dweitzman I've made changes to use vschemaacl to authorize specific users to execute "set vitess_metadata" actions. This works for our current requirements but a separate PR will separate this into its own metadataacl definition. Trying to keeps this PR small

File: go/vt/vtgate/executor_test.go (hunk 4 above)
Comment (systay) [reply go/vt/vtgate/executor_test.go]: Yeah, I just introduced it, and used it on the code I was working on at the time. You can and should definitely use it. :)

File: go/vt/topo/metadata.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+// GetMetadata retrieves all metadata value that matches the given key regular expression. If empty all values are returned.
+func (ts *Server) GetMetadata(ctx context.Context, keyFilter string) (map[string]string, error) {
+	keys, err := ts.globalCell.ListDir(ctx, MetadataPath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	keyPattern := strings.Replace(regexp.QuoteMeta(keyFilter), "%", ".*", -1)
+	keyPattern = fmt.Sprintf("^%s$", keyPattern)
+	re := regexp.MustCompile(keyPattern) // Can never fail
```
Comment (demmer) [inline go/vt/topo/metadata.go]: Are we 100% sure this is a compatible conversion from the LIKE syntax to regexp? Itt's fine for this use case but if there's a more robust and generic way to do this, we might get some mileage out of making it a common function in the Vitess code base.

Comment (tirsen) [reply go/vt/topo/metadata.go]: I found this link with examples of the LIKE syntax: https://www.w3schools.com/sql/sql_like.asp

I we also need to support `_`.

I agree that extracting a "LIKE to regexp" converter into a separate file and a more shared place and then writing specific unit tests for that would be nice. I'm not sure we need to make that 100% compliant in this PR though. We can iterate on that once we have it in a central place with a nice suite of unit tests.

Comment (demmer) [reply go/vt/topo/metadata.go]: Yep -- totally agree 100%

File: go/vt/vtgate/executor_test.go (hunk 4 above)
Comment (kalfonso) [reply go/vt/vtgate/executor_test.go]: Done. Used in the new tests

File: go/vt/topo/metadata.go (hunk 10 above)
Comment (kalfonso) [reply go/vt/topo/metadata.go]: Extracted functionality to separate file with unit tests. Also supporting '_' character and escaping using "\"

File: go/vt/topo/server.go [hunk 11]
```diff
@@ -84,6 +84,7 @@ const (
 	KeyspacesPath    = "keyspaces"
 	ShardsPath       = "shards"
 	TabletsPath      = "tablets"
+	MetadataPath     = "vitess_metadata"
```
[the remark this answers is not in the document]
Comment (tirsen) [reply go/vt/topo/server.go]: It's not too late. This is the path in zk right? Yes I agree this should be just `metadata` not `vitess_metadata`. The syntax should be `SET @@vitess_metadata` though just like all of our other SQL extensions. Right?

[the remark this answers is not in the document]
Comment (kalfonso) [reply go/vt/topo/server.go]: Renamed in this PR: https://github.com/vitessio/vitess/pull/5141

**Pull request merged**